File: generate_permutations.java
import java.util.ArrayList;
import java.util.List;

public class StringPermutations {
    
    public static List<String> generatePermutations(String s) {
        // Базовый случай
        if (s.length() <= 1) {
            List<String> result = new ArrayList<>();
            result.add(s);
            return result;
        }
        
        List<String> result = new ArrayList<>();
        
        // Перебираем каждый символ
        for (int i = 0; i < s.length(); i++) {
            // Текущий символ
            char currentChar = s.charAt(i);
            
            // Формируем оставшуюся строку
            String remaining = s.substring(0, i) + s.substring(i + 1);
            
            // Рекурсивно генерируем перестановки
            for (String perm : generatePermutations(remaining)) {
                result.add(currentChar + perm);
            }
        }
        
        return result;
    }
    // Пример использования
    public static void main(String[] args) {
        String input = "АБВ";
        List<String> permutations = generatePermutations(input);
        
        System.out.println("Все перестановки строки '" + input + "':");
        for (String p : permutations) {
            System.out.println(p);
          // Вывод:
          // Все перестановки строки 'АБВ':
          // АБВ
          // АВБ
          // БАВ
          // БВА
          // ВАБ
          // ВБА
        }
    }
}

File: generate_permutations.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

vector<string> generatePermutations(const string& s) {
    // Базовый случай
    if (s.length() <= 1) {
        return {s};
    }
    
    vector<string> result;
    
    // Перебираем каждый символ
    for (size_t i = 0; i < s.length(); ++i) {
        // Текущий символ
        char currentChar = s[i];
        
        // Формируем оставшуюся строку
        string remaining = s.substr(0, i) + s.substr(i + 1);
        
        // Рекурсивно генерируем перестановки
        vector<string> permutations = generatePermutations(remaining);
        
        // Добавляем текущий символ к каждой перестановке
        for (const string& perm : permutations) {
            result.push_back(currentChar + perm);
        }
    }
    
    return result;
}

// 1 в
int main() {
    string input = "123";
    vector<string> permutations = generatePermutations(input);
    
    cout << "Все перестановки строки '" << input << "':" << endl;
    for (const string& p : permutations) {
        cout << p << endl;
    }
    
    return 0;
}
// Вывод:
// Все перестановки строки '123':
// 123
// 132
// 213
// 231
// 312
// 321
